ang/civicase/PanelQuery.js

```diff
@@ -0,0 +1,196 @@
+      },
+      transclude: {
+        actions: '?panelQueryActions',
+        results: 'panelQueryResults'
+      }
+    };
+
+    function linkFn ($scope, $element, $attrs, $controller, $transclude) {
+      $transclude($scope, function (clone, scope) {
+        $element.find('[ng-transclude="actions"]').html(clone);
+      }, false, 'actions');
+
+      $transclude($scope, function (clone, scope) {
+        $element.find('[ng-transclude="results"]').html(clone);
+      }, false, 'results');
+    }
+  });
+
+  module.controller('panelQueryCtrl', panelQueryCtrl);
+
+  panelQueryCtrl.$inject = ['$log', '$q', '$scope', 'crmApi'];
+
+  function panelQueryCtrl ($log, $q, $scope, crmApi) {
+    var PAGE_SIZE = 5;
+
+    $scope.customData = $scope.customData || {};
+    $scope.handlers = $scope.handlers || {};
+    $scope.loading = false;
+    $scope.results = [];
+    $scope.title = '---';
+    $scope.total = 0;
+    $scope.selectedRange = 'week';
+    $scope.periodRange = [
+      { label: 'This Week', value: 'week' },
+      { label: 'This Month', value: 'month' }
+    ];
+    $scope.pagination = {
+      page: 1,
+      size: PAGE_SIZE,
+      range: { from: 1, to: PAGE_SIZE }
+    };
+
+    (function init () {
+      initWatchers();
+      verifyData();
+
+      loadData();
+    }());
+
+    /**
+     * Calculates the offset of the results list based
+     * on the current page number and page size
+     *
+     * @return {String}
+     */
+    function calculatePageOffset () {
+      return ($scope.pagination.page - 1) * $scope.pagination.size;
+    }
+
+    /**
+     * Fetches the data via the Civi API and stores the response
+     *
+     * @param {Boolean} skipCount if true then the "getcount" request won't be sent
+     * @return {Promise}
+     */
+    function fetchDataViaApi (skipCount) {
+      var apiCalls = {
+        get: [ $scope.query.entity, $scope.query.action || 'get', prepareRequestParams() ],
+        count: [ $scope.query.entity, 'getcount', $scope.query.params ]
+      };
+
+      skipCount && (delete apiCalls.count);
+
+      return crmApi(apiCalls)
+        .then(function (result) {
+          !skipCount && ($scope.total = result.count);
+
+          return processResults(result.get.values);
+        })
+        .then(function (processed) {
+          $scope.results = processed;
+        });
+    }
+
+    /**
+     * Initializes the directive's watchers
+     */
+    function initWatchers () {
+      // Triggers a refresh when the query params change
+      $scope.$watchCollection('query.params', function (newParams, oldParams) {
+        (newParams !== oldParams) && loadData();
+      }, true);
+
+      // Triggers the range handler (if present) when the selected range changes
+      $scope.$watch('selectedRange', function (newRange, oldRange) {
+        if (newRange !== oldRange && $scope.handlers.range) {
+          $scope.handlers.range($scope.selectedRange, $scope.query.params);
+        }
+      });
```

Comment:
Yes this is definitely a better solution, i think unfortunately it's just a matter of available time. I had to pick my battles in terms of what to make customizable and what to "hard code" into the directive. Given that all the dashboard panels have the range period selector, for now having it as a hardcoded part of the directive still is acceptable and of course it can be improved later on if this kind of panel needs to be reused somewhere else without the selector